engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -461,4 +262,270 @@ Type getType() {
+      if (sourceMap.containsKey(fromType)) {
+        throw new RedefinedConversionException(toType.getName(), fromType.getName());
+      } else {
+        sourceMap.put(fromType, function);
+      }
+    }
+
+    /**
+     * Registers a new symbol in the polyglot namespace.
+     *
+     * @param name the name of the symbol
+     * @param sym the value being exposed
+     */
+    public void registerPolyglotSymbol(String name, Object sym) {
+      assert moduleScope == null;
+      polyglotSymbols.put(name, sym);
+    }
+
+    /**
+     * Registers all methods of a type in the provided scope.
+     *
+     * @param tpe the methods of which type should be registered
+     * @param scope target scope where methods should be registered to
+     */
+    public void registerAllMethodsOfTypeToScope(Type tpe, ModuleScope.Builder scope) {
+      // FIXME: because of Builtins can't enable 'assert moduleScope == null;'
+      Type tpeKey = tpe == null ? noTypeKey : tpe;
+      var allTypeMethods = methods.get(tpeKey);
+      if (allTypeMethods != null) {
+        allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpeKey, name, fun));
+      }
+    }
+
+    /**
+     * Adds a dependency for this module.
+     *
+     * @param scope the scope of the newly added dependency
+     */
+    public void addImport(ModuleScope.Builder scope) {
+      assert moduleScope == null;
+      imports.add(scope);
+    }
+
+    /**
+     * Adds an information about the module exporting another module.
+     *
+     * @param scope the exported scope
+     */
+    public void addExport(ModuleScope.Builder scope) {
+      assert moduleScope == null;
+      exports.add(scope);
+    }
+
+    public Module getModule() {
+      return module;
+    }
+
+    public Type getType(String typeName) {
+      // assert moduleScope == null;
+      return types.get(typeName);
+    }
+
+    /**
+     * @return the associated type of this module.
+     */
+    public Type getAssociatedType() {
+      return associatedType;
+    }
+
+    public Object getPolyglotSymbol(String symbolName) {
+      return polyglotSymbols.get(symbolName);
+    }
+
+    /**
+     * Create a copy of this `ModuleScope` while taking into account only the provided list of
+     * types.
+     *
+     * @param typeNames list of types to copy to the new scope
+     * @return a copy of this scope modulo the requested types
+     */
+    public ModuleScope.Builder withTypes(List<String> typeNames) {
+      Map<String, Object> polyglotSymbols = new HashMap<>(this.polyglotSymbols);
+      Map<String, Type> requestedTypes = new HashMap<>(this.types);
+      Map<Type, Map<String, Supplier<Function>>> methods = new ConcurrentHashMap<>();
+      Map<Type, Map<Type, Function>> conversions = new ConcurrentHashMap<>();
+      Set<ModuleScope.Builder> imports = new HashSet<>(this.imports);
+      Set<ModuleScope.Builder> exports = new HashSet<>(this.exports);
+      this.types
+          .entrySet()
```

Comment:
```suggestion
          .forEach((key, value) -> {
            if (typeNames.contains(key)) {
              requestedTypes.put(key, value);
            }
          });
```

Reply:
`for (var en : types.entrySet()` is easier to step thru in debugger.

Reply:
When you apply my suggestion, you don't have to call `entry.getKey()` and `entry.getValue()` in the subsequent code. That was the main reason for the suggestion.